src/cargo/core/compiler/custom_build.rs

```diff
@@ -679,46 +687,90 @@ impl BuildOutput {
         let mut warnings = Vec::new();
         let whence = format!("build script of `{}`", pkg_descr);
 
+        // Old syntax:
+        //    cargo:rustc-flags=VALUE
+        //    cargo:KEY=VALUE (for other unreserved keys)
+        // New syntax:
+        //    cargo::rustc-flags=VALUE
+        //    cargo::metadata=KEY=VALUE (for other unreserved keys)
+        // Due to backwards compatibility, no new keys can be added to this old format.
+        const RESERVED_PREFIXES: &[&str] = &[
+            "rustc-flags=",
+            "rustc-link-lib=",
+            "rustc-link-search=",
+            "rustc-link-arg-cdylib=",
+            "rustc-link-arg-bins=",
+            "rustc-link-arg-bin=",
+            "rustc-link-arg-tests=",
+            "rustc-link-arg-benches=",
+            "rustc-link-arg-examples=",
+            "rustc-link-arg=",
+            "rustc-cfg=",
+            "rustc-check-cfg=",
+            "rustc-env=",
+            "warning=",
+            "rerun-if-changed=",
+            "rerun-if-env-changed=",
+        ];
+        const DOCS_LINK_SUGGESTION: &str = "See https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script \
+                for more information about build script outputs.";
+
+        fn parse_directive<'a>(
+            whence: &str,
+            line: &str,
+            data: &'a str,
+        ) -> CargoResult<(&'a str, &'a str)> {
+            let mut iter = data.splitn(2, "=");
+            let key = iter.next();
+            let value = iter.next();
+            match (key, value) {
+                (Some(a), Some(b)) => Ok((a, b.trim_end())),
+                _ => bail!(
+                    "invalid output in {}: `{}`\n\
+                    Expected a line with `cargo::KEY=VALUE` with an `=` character, \
+                    but none was found.\n\
+                    {}",
+                    whence,
+                    line,
+                    DOCS_LINK_SUGGESTION
+                ),
+            }
+        }
+
         for line in input.split(|b| *b == b'\n') {
             let line = match str::from_utf8(line) {
                 Ok(line) => line.trim(),
                 Err(..) => continue,
             };
-            let mut iter = line.splitn(2, ':');
-            if iter.next() != Some("cargo") {
-                // skip this line since it doesn't start with "cargo:"
+            let (key, value) = if let Some(data) = line.strip_prefix("cargo::") {
+                // For instance, `cargo::rustc-flags=foo` or `cargo::metadata=foo=bar`.
+                parse_directive(whence.as_str(), line, data)?
+            } else if let Some(data) = line.strip_prefix("cargo:") {
```

Comment:
@ehuss
> I'm wondering if it would help to reserve cargo:: as an error for now, and then in around 6 months stabilize this change? This could provide better error messages for the cases mentioned above.

Would it work for us:
- Put a `if !nightly { error }` in here
- Keep the issue open and put in a reminder to revisit this in 6 months to remove the error and close the issue?

Reply:
I would prefer to generally avoid nightly-specific behavior changes that aren't opt-in. It has the same risks as mentioned above where a user may start using it (and they are using nightly) and not realize that it is nightly-only.

Reply:
@ehuss Do you mean we need to bail out an error when using `cargo::` and wait for 6 months, then we can add this new syntax?

Reply:
> * Put a `if !nightly { error }` in here

Do we need to be precise to a version? I understand that nightly before this pr doesn't work with it either.

Reply:
> Do you mean we need to bail out an error when using cargo:: and wait for 6 months, then we can add this new syntax?

Yea, that is what I'm proposing.

Reply:
Sounds good to me. What about adding an unstable flag for the new syntax? Then people will only use it as an unstable feature. Does this help to avoid misuse in the old version Cargo?

Reply:
I think it would be fine to add an unstable flag. That would at least keep this code in a ready state to stabilize. However, I wouldn't expect anyone to actually use it since by itself it doesn't provide any benefit. We're only making this change to make it safer to add new instructions in the future.

Reply:
@epage What do you think? Should I add an unstable flag or add an unstable flag?

Reply:
I leave it to you as for what works best for you.   There are benefits and costs either direction
-  I see benefit in keeping the core of this code in a running state as we continue to make changes
  - However, that leaves off documentation and test updates which is where I'm assuming the bulk of the change lives (and tests would likely be the bulk of the divergence)
  - We then likely would also add tests around the feature flag which is more work
  - We have to duplicate (and rename) build script tests instead of updating in place
- By merging with a flag now, people can build on this to to add new directives which will then be ready once this change stabilizes (like the `error` directive requested in #10159)
- For at least some contributors, it feels more satisfactory to merge a change now than to wait later

Reply:
> * However, that leaves off documentation and test updates which is where I'm assuming the bulk of the change lives (and tests would likely be the bulk of the divergence)
> * We then likely would also add tests around the feature flag which is more work
> * We have to duplicate (and rename) build script tests instead of updating in place

Yes. It involves a lot of changes. And the core changes are pretty simple and small. If we need to leave these documents and test updates now. Then I think I would choose to report the error first, because even if it's an unstable feature we still need to update these documents and tests in the future. Adding an unstable flag would be a lot more work and not very rewarding. I'm sure I'll still be contributing to cargo in 6 months' time, so waiting is fine with me.

Thanks for your review!  💚 💙 💜 💛 ❤️